src/platforms/rpi-vc4/display.cpp

```diff
@@ -0,0 +1,357 @@
+        output.type = mg::DisplayConfigurationOutputType::hdmia;
+        output.pixel_formats = std::vector<MirPixelFormat>{input_format_to_pixel_format(info.input_format)};
+        output.modes = std::vector<mg::DisplayConfigurationMode>{mg::DisplayConfigurationMode{
+            geometry::Size{info.width, info.height},
+            0  // Um, we don't get this info!
+        }};
+        output.preferred_mode_index = 0;
+        output.connected = true;
+        output.current_mode_index = 0;
+        output.current_format = output.pixel_formats[0];
+        output.power_mode = MirPowerMode::mir_power_mode_on;
+        // TODO: orientation from mode.transform
+        output.gamma_supported = MirOutputGammaSupported::mir_output_gamma_unsupported;
+        output.form_factor = MirFormFactor::mir_form_factor_unknown;
+	output.orientation = mir_orientation_normal;
+	output.scale = 1.0f;
+	output.subpixel_arrangement = mir_subpixel_arrangement_unknown;
+    output.used = true;
+
+        return output;
+    }
+
+    static auto input_format_to_pixel_format(DISPLAY_INPUT_FORMAT_T format) -> MirPixelFormat
+    {
+        switch (format)
+        {
+        case DISPLAY_INPUT_FORMAT_RGB888:
+            return mir_pixel_format_rgb_888;
+        case DISPLAY_INPUT_FORMAT_RGB565:
+            return mir_pixel_format_rgb_565;
+        case DISPLAY_INPUT_FORMAT_INVALID:
+            return mir_pixel_format_rgb_888;
+//            BOOST_THROW_EXCEPTION((std::runtime_error{"Display has invalid pixel format?"}));
+        }
+#ifndef __clang__
+        BOOST_THROW_EXCEPTION((std::logic_error{"Unhandled DISPLAY_INPUT_FORMAT type!"}));
+#endif
+    }
+
+    explicit DisplayConfiguration(mg::DisplayConfigurationOutput from)
+        : the_output{std::move(from)}
+    {
+    }
+
+    mg::DisplayConfigurationOutput the_output;
+};
+
+namespace
+{
+EGLConfig choose_config(EGLDisplay display, mg::GLConfig const& requested_config)
+{
+    EGLint const config_attr[] = {
+        EGL_SURFACE_TYPE,       EGL_WINDOW_BIT | EGL_PBUFFER_BIT,
+        EGL_RED_SIZE,           8,
+        EGL_GREEN_SIZE,         8,
+        EGL_BLUE_SIZE,          8,
+        EGL_ALPHA_SIZE,         0,
+        EGL_DEPTH_SIZE,         requested_config.depth_buffer_bits(),
+        EGL_STENCIL_SIZE,       requested_config.stencil_buffer_bits(),
+        EGL_RENDERABLE_TYPE,    MIR_SERVER_EGL_OPENGL_BIT,
+        EGL_NONE};
+
+    EGLint num_egl_configs;
+    EGLConfig egl_config;
+    if (eglChooseConfig(display, config_attr, &egl_config, 1, &num_egl_configs) != EGL_TRUE)
+    {
+        BOOST_THROW_EXCEPTION(mg::egl_error("Failed to chose EGL config"));
+    }
+    else if (num_egl_configs != 1)
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error{"Failed to find compatible EGL config"});
+    }
+
+    return egl_config;
+}
+
+EGLContext create_context(EGLDisplay display, EGLConfig config)
+{
+    eglBindAPI(MIR_SERVER_EGL_OPENGL_API);
+
+    EGLint const context_attr[] = {
```

Comment:
I see this definition repeated in the next function, could it be moved to namespace scope and shared?